server/src/ankaios_server/update_state.rs

```diff
@@ -1,438 +0,0 @@
-    // [impl->swdd~update-current-state-empty-update-mask~1]
-    if update.update_mask.is_empty() {
-        return Ok(update.state);
-    }
-
-    // [impl->swdd~update-current-state-with-update-mask~1]
-    let mut new_state: Object = current_state.try_into().map_err(|err| {
-        UpdateStateError::ResultInvalid(format!("Failed to parse current state, '{}'", err))
-    })?;
-    let state_from_update: Object = update.state.try_into().map_err(|err| {
-        UpdateStateError::ResultInvalid(format!("Failed to parse new state, '{}'", err))
-    })?;
-
-    for field in update.update_mask {
-        let field: Path = field.into();
-        if let Some(field_from_update) = state_from_update.get(&field) {
-            if new_state.set(&field, field_from_update.to_owned()).is_err() {
-                return Err(UpdateStateError::FieldNotFound(field.into()));
-            }
-        } else if new_state.remove(&field).is_err() {
-            return Err(UpdateStateError::FieldNotFound(field.into()));
-        }
-    }
-
-    if let Ok(new_state) = new_state.try_into() {
-        Ok(new_state)
-    } else {
-        Err(UpdateStateError::ResultInvalid(
-            "Could not parse into CompleteState.".to_string(),
-        ))
-    }
-}
-
-pub fn prepare_update_workload(
-    current_state: &State,
-    new_state: &State,
-) -> Option<common::execution_interface::ExecutionCommand> {
-    let mut added_workloads: Vec<WorkloadSpec> = Vec::new();
-    let mut deleted_workloads: Vec<DeletedWorkload> = Vec::new();
-
-    // find updated or deleted workloads
-    current_state.workloads.iter().for_each(|(wl_name, wls)| {
-        if let Some(new_wls) = new_state.workloads.get(wl_name) {
-            // The new workload is identical with existing or updated. Lets check if it is an update.
-            if wls != new_wls {
-                // [impl->swdd~server-detects-changed-workload~1]
-                added_workloads.push(new_wls.clone());
-                deleted_workloads.push(DeletedWorkload {
-                    agent: wls.agent.clone(),
-                    name: wl_name.clone(),
-                    dependencies: HashMap::new(),
-                });
-            }
-        } else {
-            // [impl->swdd~server-detects-deleted-workload~1]
-            deleted_workloads.push(DeletedWorkload {
-                agent: wls.agent.clone(),
-                name: wl_name.clone(),
-                dependencies: HashMap::new(),
-            });
-        }
-    });
-
-    // find new workloads
-    // [impl->swdd~server-detects-new-workload~1]
-    new_state
-        .workloads
-        .iter()
-        .for_each(|(new_wl_name, new_wls)| {
-            if !current_state.workloads.contains_key(new_wl_name) {
-                added_workloads.push(new_wls.clone());
-            }
-        });
-
-    if added_workloads.is_empty() && deleted_workloads.is_empty() {
-        return None;
-    }
-
-    log::info!(
```

Comment:
Is this trace available somewhere in the new code? It is rather a matter of taste, but I prefer to have an info trace saying something about content of the update (and therefore about what the server is supposed to do).

Reply:
Yes in general I agree, the reason why I have removed the log was because of effort/value estimation. If I keep this log, then the log is also printed on startup state check as well, which feels strange if it contains a cycle and the cycle error message comes after it. If I want to print this only on subsequent updates, then I need to duplicate code or I must put another "let else" to extract the containers out of the enum again, which I personally did not like (then you need unreachable once more...). I will check what I can do else.

Reply:
Hmm I have kept like before the changes that the "update_state" method returns an execution request. But, I think I can return the containers somehow, but not inside an enum because it feels unhandy here and I think an update_state method should anyway not return a command object, which then another object sends. Let me try something out, I will see how I can change this.

Reply:
I have refactored and checked how it looks like with an tuple as return type and the extra step to extract together with the unreachable statement is even omitted. And we can have back the log only for subsequent update. However, I have now seen that I can directly call the cycle check in the start method, because I know how many workloads were added (all inside the first state). But the downside is that I need something like an init or setter in the server state which I wanted to avoid in general because it potentially bypasses some in-variants of the server state or looks weird. But I will discuss that with @krucod3 tomorrow.

Reply:
It is also fine to keep it without a trace. Do not make it too complicated.

Reply:
I have found a solution which has not too much effort. I will commit it tomorrow morning.

Reply:
I have commit the solution of changing the return type. Now the log is output again if the checks on the new state have passed.

Reply:
The trace looks good 👍